File: ts2mls/js/src/test/scala/Tuple.scala
import org.scalatest.funsuite.AnyFunSuite
import ts2mls.TSProgram
import ts2mls.types._

class Tuple extends AnyFunSuite {
  test("Tuple") {
    val program = TSProgram(Seq("src/test/typescript/Tuple.ts"))
    assert(TypeCompare(program.>("key"), "([string, boolean]) => string"))
    assert(TypeCompare(program.>("value"), "([string, boolean]) => boolean"))
    assert(TypeCompare(program.>("third"), "([number, number, number]) => number"))
    assert(TypeCompare(program.>("vec2"), "(number, number) => [number, number]"))
    assert(TypeCompare(program.>("twoFunctions"), "([(number) => number, (number) => number], number) => number"))
    assert(TypeCompare(program.>("tupleIt"), "(string) => [string]"))
    assert(TypeCompare(program.>("s"), "(boolean) => [string | number, number | false | true]"))
    assert(TypeCompare(program.>("s2"), "([boolean, string | number]) => string | number"))
    assert(TypeCompare(program.>("ex"), "(T', U') => [T', U', T' & U']"))
    assert(TypeCompare(program.>("foo"), "([T' & U']) => void"))
    assert(TypeCompare(program.>("conv"), "({y: number}) => [{y: number}, {z: string}]"))
    assert(TypeCompare(program.>("swap"), "([{x: number}, {}]) => [B, A]"))
    assert(TypeCompare(program.>("fff"), "(FFF<string>, string) => void"))
    assert(TypeCompare(program.>("getFFF"), "FFF<number>"))
  }

  test("Tuple Convert") {
    import mlscript._

    val program = TSProgram(Seq("src/test/typescript/Tuple.ts"))

    program.getMLSType("key") match {
      case Function(p, r) => p match {
        case mlscript.Tuple(lst) => {
          lst(0) match {
            case None -> Field(in, out) => {
              assert(in.isEmpty)
              out match {
                case TypeName(name) => assert(name.equals("string"))
                case _ => assert(false)
              }
            }
          }

          lst(1) match {
            case None -> Field(in, out) => {
              assert(in.isEmpty)
              out match {
                case TypeName(name) => assert(name.equals("bool"))
                case _ => assert(false)
              }
            }
          }
        }
        case _ => assert(false)
      } 
      case _ => assert(false)
    }
  }
}

File: ts2mls/js/src/test/typescript/Tuple.ts
function key(x: [string, boolean]): string {
  return x[0];
}

function value(x: [string, boolean]): boolean {
  return x[1];
}

function third(x: [number, number, number]): number {
  return x[2];
}

function vec2(x: number, y: number): [number, number] {
  return [x, y];
}

function twoFunctions(ff: [(x: number) => number, (x: number) => number], x: number): number {
  return ff[0](x) + ff[1](x);
}

function tupleIt(x: string): [() => string] {
  return [function() { return x }]
}

function s(flag: boolean): [string | number, number | boolean] {
  if (flag) {
      return ["abc", 12];
  }
  else {
      return [24, false];
  }
}

function s2(t: [boolean, string | number]): string | number {
  if (t[0]) return t[1]
  else return 0
}

function ex<T, U>(x: T, y: U): [T, U, T & U] {
  return [x, y , <T & U>{}];
}

function foo<T, U>(x: [T & U]) {}

function conv(x: {y: number}): [{y: number}, {z: string}] {
  return [x, {z: x.y.toString()}];
}

class A {
  x: number
}
class B {}

function swap(x: [A, B]): [B, A] {
  return [x[1], x[0]];
}